Exception consistency (pt.1). Stateless schema validator

scaytrase opened the pull request:
Key changes:
* Validator instance is now statless. It can be re-used to validate other data against other schema.
* Consistent `\Schema\Exception`
  * new exception class - InvalidSchema - for converting defensive respect validations. It's considered Runtime, so not reflected in sigatures as defencive check
  * new exception class - SchemaMismatch - as a base class for all other validation exceptions
  * new exception hierarchy
     * SchemaMismatch
         * supports breadcrumbs on this level
     * ValidationKeywordFailed (I find this naming strange, but didn't touch it)
     * TypeMismatch (keyword === type)
     * FormatMismatch (keyword === type)
* Validator interface


Also I've updated all signatures in `\Schema` namespace to reflect real exceptions that are expected as validation result.

Fixes #14, partially #18

scaytrase committed:
Exception consistency. Stateless validator

scaytrase commented on src/PSR7/Validators/QueryArguments.php:
```diff
@@ -26,18 +29,19 @@ public function validate(MessageInterface $message, array $specs) : void
         // Check if message misses query argument
         foreach ($specs as $name => $spec) {
-            if (! array_key_exists($name, $message->getQueryParams()) && $spec->required) {
+            if ($spec->required && ! array_key_exists($name, $message->getQueryParams())) {
```
just a microoptimisation

lezhnev74 replied:
Good catch!

scaytrase commented on src/Schema/Keywords/MultipleOf.php:
```diff
@@ -18,18 +18,21 @@ class MultipleOf extends BaseKeyword
-            if ($data % $multipleOf) {
-                throw new Exception(sprintf('Division by %d did not resulted in integer', $multipleOf));
+            $value = $data / $multipleOf;
+            if ((int) $value !== $value) {
```
https://swagger.io/docs/specification/data-models/data-types/#multipleOf

`multipleOf` could be used against floats, so I patched this a bit while inspecting validator

scaytrase commented on src/Schema/Keywords/Not.php:
```diff
@@ -35,23 +35,25 @@ public function __construct(CebeSchema $parentSchema, int $type, BreadCrumb $bre
+            try {
+                $schemaValidator->validate($data, $not, $this->dataBreadCrumb);
+            } catch (SchemaMismatch $e) {
+                return;
```
Simplified flow here a bit

lezhnev74 replied:
Looks good!

scaytrase commented on src/Schema/Keywords/OneOf.php:
```diff
@@ -47,23 +50,22 @@ public function validate($data, array $oneOf) : void
             // Validate against all schemas
-            $matchedCount = 0;
+            $matchedCount    = 0;
+            $schemaValidator = new SchemaValidator($this->validationDataType);
```
With stateless `SchemaValidator` we can instantiate it once

lezhnev74 replied:
Good for performance, indeed.

scaytrase commented on src/Schema/Keywords/Required.php:
```diff
@@ -52,23 +55,27 @@ public function validate($data, array $required) : void
                     if ((
-                            $this->parentSchema->properties[$reqProperty]->writeOnly &&
-                            $this->validationDataType === \OpenAPIValidation\Schema\Validator::VALIDATE_AS_RESPONSE
+                            ($this->parentSchema->properties[$reqProperty]->writeOnly ?? false) &&
```
small inline bugfix for unset fields or parent properties

lezhnev74 replied:
Good catch. That would have troubled us in the future.

scaytrase commented on src/Schema/Keywords/Type.php:
```diff
@@ -33,70 +32,68 @@ class Type extends BaseKeyword
+                }
+                break;
+            default:
+                throw InvalidSchema::becauseTypeIsNotKnown($type);
```
Now it's also invalid schema situation, so changed the exception

lezhnev74 replied:
Makes sense

scaytrase commented on src/Schema/Keywords/UniqueItems.php:
```diff
@@ -25,17 +24,23 @@ class UniqueItems extends BaseKeyword
      */
     public function validate($data, bool $uniqueItems) : void
     {
+        if (! $uniqueItems) {
```
Also small bugfix. `uniqueItems` could be set to `false` explicitly

lezhnev74 replied:
Excellent catch!

scaytrase commented on src/Schema/Validator.php:
```diff
@@ -4,173 +4,15 @@
-    {
-        return $this->dataBreadCrumb ?? new BreadCrumb();
-    }
+    public function validate($data, Schema $schema, ?BreadCrumb $breadCrumb = null) : void;
```
As a future idea i suggest to replace third argument to some kind of `ValidationContext` instead (not in this PR). WDYT?

lezhnev74 replied:
In general, yes, I support this idea. 
That would be a good thing when(if) there would be more arguments to pass in. 
If there is just one thing at the moment (a bread crumb), that looks ok-ish to me.

scaytrase replied:
Yea, thats why I call it `future` idea ) Not now for sure

scaytrase commented on tests/Schema/Keywords/TypeTest.php:
```diff
@@ -62,11 +62,8 @@ public function testItValidatesTypeRed() : void
-            } catch (ValidationKeywordFailed $e) {
-                $this->assertEquals('type', $e->keyword());
-            }
+            $this->expectException(TypeMismatch::class);
```
Some tests becase a bit more clear, where exception inspection is not needed

lezhnev74 replied:
Yeah, looks good

scaytrase commented on src/Schema/Keywords/MultipleOf.php:
```diff
@@ -18,18 +18,21 @@ class MultipleOf extends BaseKeyword
-            if ($data % $multipleOf) {
-                throw new Exception(sprintf('Division by %d did not resulted in integer', $multipleOf));
+            $value = $data / $multipleOf;
+            if ((int) $value != $value) {
```
We have a problem here though. `phpcs` requires strict comparison. I'll think how to fix it

scaytrase commented on src/Schema/Keywords/AnyOf.php:
```diff
@@ -39,31 +39,29 @@ public function __construct(CebeSchema $parentSchema, int $type, BreadCrumb $bre
-                } catch (ValidationKeywordFailed $e) {
+                    $schemaValidator->validate($data, $schema, $this->dataBreadCrumb);
+
+                    return;
```
Simplified flow. We do not really need to continue here if we match something

lezhnev74 replied:
Another great performance related catch!

lezhnev74 commented:
@scaytrase I need to collaborate with you on this PR. Is it possible to have write access to the repo, so I can just commit directly here?

scaytrase commented:
I hope checkbox `allow edits from maintainers` has already granted you this acess. I think you've already fixed my PRs before by adding commits there. But if you need more permissions I'll give you them for sure

lezhnev74 committed:
Add failing tests for multipleOf validator

These tests fail for particular data combination. Need review of implementation.

lezhnev74 commented on tests/Schema/Keywords/MultipleOfTest.php:
```diff
@@ -5,39 +5,76 @@
+use OpenAPIValidation\Schema\SchemaValidator;
 use OpenAPIValidationTests\Schema\SchemaValidatorTest;
 
 final class MultipleOfTest extends SchemaValidatorTest
```
Please review this test. It fails at the moment.

scaytrase replied:
Weird, but fixed. int/int is int if it's divisible, not float. what the hell. added explicit cast to float for comparison

lezhnev74 replied:
As it turns out, the result of a division can be both `int` and `float`. It would be better to have `float` as a result of a division in all cases (thought PHP is not there yet).
```php
4/2 === 2; // true
4.0/2 === 2.0; //true
```

scaytrase replied:
yeah, that's the thing I did in the fix. Updated it to be more clear

lezhnev74 committed:
Add extra comments to explain how  Required validator works

lezhnev74 commented:
> ValidationKeywordFailed (I find this naming strange, but didn't touch it)

Please suggest a better name here. 
My intention was explaining exactly what happened - a schema keyword (which is an OpenAPI term(JSON schema)) failed validation against data.

scaytrase committed:
Fix multipleOf test

scaytrase commented:
> Please suggest a better name here.

Well, I'd better start with `KeywordValidationFailed`. But there is no keywod validation, there is data validation. To keep naming style we could use something like `KeywordMismatch` or so, but not sure

lezhnev74 commented:
`KeywordMismatch` looks like a good option. 

There could be other exceptions which extend this one (if we want to report some specific keyword related failure (hence `FormatMismatch -> TypeMismatch -> ValidationKeywordFailed`).

scaytrase committed:
Rename ValidationKeywordFailed to KeywordMismatch

scaytrase committed:
Make multipleOf check more clear

lezhnev74 commented:
I am going to study changes for a day or two, maybe. If nothing major found, then gonna tag `0.13.0`.

lezhnev74 committed:
Update readme

lezhnev74 committed:
Update Readme

lezhnev74 committed:
Update Readme

lezhnev74 committed:
Update keyword validators

Decouple input validation from keyword validation

lezhnev74 committed:
Add comment to InvalidSchema

lezhnev74 commented:
@scaytrase please review this commit: https://github.com/lezhnev74/openapi-psr7-validator/pull/24/commits/ecfda741315fa1eb2af72f3c5d21be9767d4ab9f

I thought that input validation must be decoupled from "keyword" validation. But now I have doubts as it sometimes calls `SchemaValidator` like here: https://github.com/lezhnev74/openapi-psr7-validator/pull/24/commits/ecfda741315fa1eb2af72f3c5d21be9767d4ab9f#diff-e724b0820a8b3411744e47d319e09430

scaytrase commented:
> I thought that input validation must be decoupled from "keyword" validation. 

Looks totally valid. I thought about same change but didn't  do it for no reason. As long as `SchemaValidator` keeps it contract defensive validation could be decoupled, as `Respect` exceptions will not pass thru the catch clause and contracts would stay valid and clear

lezhnev74 commented on src/Schema/Validator.php:
```diff
@@ -4,173 +4,15 @@
 
-// This will load a whole schema and data to validate if one matches another
-class Validator
+interface Validator
```
Can you explain the logic behind extracting this interface from the `SchemaValidator`? Looks like it is not used anywhere.

scaytrase replied:
Well, at least it can be used as a library public contract, allowing to mock it easily. I have other ideas but not sure about them

lezhnev74 replied:
It does not look like the implementation can be swapped with something else. I don't mind this interface, I wonder what are the other possible cases?

scaytrase replied:
well, it could be mocked and decorated using the interface. currently there is no points of extension for this class, so even we definalize it back in order to mock or decorate it we will get only FQCN and method signature. It's called interface, so I extracted it explicitly.

My other thoughts are about looking at other nested validator. Most of them really have the same signature, but a bit smudged. In general they take the a portion of `data` and the `schema` in some way and compare one with another. So signature (inlc. exceptions) stays the same.

If we incapsulate keyword knowledge into the keyword validator itself (like moving `isset` check inside it) and pass not keyword options but schema object - we will create incapsulated validators with the same signature, which has all knowled about data validation

lezhnev74 replied:
Sounds good to me

scaytrase commented:
I'll point to PSR-7 namespace consistency a bit later. I hope there will be similar changes.
